staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -584,7 +584,7 @@ func (c *Cacher) Get(ctx context.Context, key string, resourceVersion string, ob
 func (c *Cacher) GetToList(ctx context.Context, key string, resourceVersion string, pred storage.SelectionPredicate, listObj runtime.Object) error {
 	pagingEnabled := utilfeature.DefaultFeatureGate.Enabled(features.APIListChunking)
 	hasContinuation := pagingEnabled && len(pred.Continue) > 0
-	hasLimit := pagingEnabled && pred.Limit > 0 && resourceVersion != "0"
+	hasLimit := pagingEnabled && pred.Limit > 0 && resourceVersion == ""
```

Comment:
how does this change behavior from a <= 1.16 client's perspective?

Reply:
Basically it means that if Limit is set, only requests with RV="" will be served from etcd.
So what changes is:
- for requests with RV=value != {"", "0"} and Limit set, instead of 
  "exact RV="rv"" semantic, we get "at least RV="rv"" semantic

Which is consistent with what we document here:
https://kubernetes.io/docs/reference/using-api/api-concepts/#the-resourceversion-parameter

[and this is unrelated to client version - it impact clients in all versions the same way]

Reply:
This has the effect of no longer paging list requests like `?limit=500&resourceVersion=123`

Reply:
By serving from cache, doesn't this change behavior of `limit=n,resourceVersion={non-0-value}` from "Exact" to "Not older than"?

Reply:
> This has the effect of no longer paging list requests like ?limit=500&resourceVersion=123

Yes - that is true.
The question is who was using such requests (in k/k we started utilizing them with https://github.com/kubernetes/kubernetes/pull/83520, so in 1.17)

> By serving from cache, doesn't this also change behavior of limit=n,resourceVersion={non-0-value} from "Exact" to "Not older than"?

yeah - that's exactly what I wrote two comments above:

> for requests with RV=value != {"", "0"} and Limit set, instead of
"exact RV="rv"" semantic, we get "at least RV="rv"" semantic

Reply:
Also, since list resourceVersions can be obtained by a client via a list from etcd, and reflect the server's current version, but the resourceVersion observed by the server's storage cacher depends on having seen a resourceVersion in a watch event as part of an object change, this would make it possible for the apiserver to permanently time-out/reject list requests with valid resourceVersions

1. API server starts up, populates pod watch cache with a list at rv=10
2. Changes to pods happen at rv=11,12,13,14,15
3. API server watch cache observes those changes, sees "most recent rv" for pods as 15
4. Changes to configmaps occur at rv=16,17,18,19,20
5. Client does a list of `pods?resourceVersion=`, gets a current resource version of 20
6. Client uses that resourceVersion to do a list of `pods?limit=500&resourceVersion=20`, API server waits 3 seconds, then returns NewTooLargeResourceVersionError

It's not clear how a client in that scenario could make progress.

Reply:
Agree that it's the problem - and that is one of the reasons this is market as WIP.
I wanted to evaluate performance of that (it seems that this is causing a lot of problem).

To solve this case, what i was thinking about was to:
- wait for some short amount of time (e.g. 1s or 500ms or sth) for the RV in cache
- if we didn't make enough progress, fallback to etcd

That will solve the "inability to make progress" bug.

Reply:
The real problem here is that etcd 3.4 isn't being used.  Why don't we focus on pushing that forward?  The faster we get that better everyone gets.

Reply:
@smarterclayton - I don't fully understand it? I don't think any etcd feature and its features (including progress-notify which btw I'm trying to push: https://github.com/kubernetes/enhancements/pull/1404 won't solve this problem...)

Reply:
I think I know how to solve our problem better - basically we need to change the reflector to not set limit when we set RV!="".
But effectively that would be the same so that is easier to prove that we achieve what we want.

I will also open an issue with all the details later today to provide more context to this change.

Reply:
FTR - more info is in https://github.com/kubernetes/kubernetes/issues/86483